vignettes/writing.Rmd

```diff
@@ -0,0 +1,288 @@
+---
+title: "Writing dust2 systems"
+output: rmarkdown::html_vignette
+vignette: >
+  %\VignetteIndexEntry{Writing dust2 systems}
+  %\VignetteEngine{knitr::rmarkdown}
+  %\VignetteEncoding{UTF-8}
+---
+
+```{r, include = FALSE}
+source("support.R")
+```
+
+This vignette covers how to write `dust2` systems from scratch.  Users of this package via [`odin2`](https://mrc-ide.github.io/odin2) do not need to read or understand this vignette, though of course you are welcome to read it.  We will start with an almost comically trivial system and try to expand it over several sections.
+
+# First steps, a random walk
+
+Consider a unbiased random walk; at each time step we move our position with a draw from a normal distribution with mean 0 and some standard deviation.
+
+To implement this model in dust we write out a C++ file that looks like:
+
+```{r walk_code, echo = FALSE, results = "asis"}
+walk_code <- readLines("examples/walk.cpp")
+cc_output(walk_code)
+```
+
+All the code sits inside a class, though in this class every method is `static` (we have deleted the constructor by writing `walk() = delete;` to emphasise this).  All dust class definitions have the same components, but the order is not terribly important:
+
+* a series of annotations (the comments starting with `// [[dust2::`)
+* some type declarations and `struct` definitions
+* a series of `static` methods.  For these, the types of the arguments must match the interface described here
+
+We consider each of these pieces in turn here, and consider the *other* values that might have been used in place.
+
+## Support code
+
+We include a header from `dust2`; the path for the compiler to find this will be arranged by R using its `LinkingTo` system.  This include must be present.
+
+```{r, echo = FALSE, results = "asis"}
+i <- grep("#include", walk_code)
+stopifnot(i == 1)
+cc_output(walk_code[[i]])
+```
+
+## Annotations
+
+The first annotation tells `dust2` the name of the class that contains the system definition.  This may seem obvious but computers are actually not very clever and it's best to be blunt with them.
+
+```{r, echo = FALSE, results = "asis"}
+annotations <- walk_code[startsWith(walk_code, "// [[")]
+i <- grep("dust2::class", annotations)
+stopifnot(length(i) == 1, i == 1)
+cc_output(annotations[[i]])
+```
+
+The second annotation tells `dust2` that this is a discrete-time (as opposed to continuous time) system:
+
+```{r, echo = FALSE, results = "asis"}
+i <- grep("dust2::time_type", annotations)
+stopifnot(length(i) == 1, i == 2)
+cc_output(annotations[[i]])
+```
+
+If we had wanted a continuous time system of ODEs we would have written `// [[dust2::time_type(continuous)]]` and then later some of the rest of the code would be different; we will discuss this later.
```

Comment:
Struggling to do the backticks as a suggestion - but in my preview, there's a line-break after the `//` and before [[dust2:: which would be good to avoid - I was trying to suggest a ```r ``` block